stdlib/public/core/CString.swift

```diff
@@ -419,19 +473,32 @@ extension String {
     self = String.decodeCString(nullTerminatedCodeUnits, as: sourceEncoding)!.0
   }
 
-  @_specialize(where Encoding == Unicode.UTF8)
-  @_specialize(where Encoding == Unicode.UTF16)
-  @inlinable // Fold away specializations
+  /// Creates a new string by copying the null-terminated sequence of code units
+  /// referenced by the given array.
+  ///
+  /// If `nullTerminatedCodeUnits` contains ill-formed code unit sequences, this
+  /// initializer replaces them with the Unicode replacement character
+  /// (`"\u{FFFD}"`).
+  ///
+  /// Note: This initializer is deprecated. Use the initializer
+  ///       `String.init(decoding: array, as: Encoding.self)` instead,
+  ///       remembering that "\0" is a valid character in Swift.
+  ///
+  /// - Parameters:
+  ///   - nullTerminatedCodeUnits: An array containing a null-terminated
+  ///     sequence of code units encoded in `sourceEncoding`.
+  ///   - sourceEncoding: The encoding in which the code units should be
+  ///     interpreted.
+  @inlinable
   @_alwaysEmitIntoClient
+  @available(swift, deprecated: 6, message: "Use String(decoding: array, as: Encoding.self) instead")
```

Comment:
Out of curiosity, why are these depreciations for Swift 6? Are the replacements not added until then?

Reply:
The cleanest way to soft-deprecate is to hang the deprecation on a future language mode; Xcode de-emphasizes the deprecated spelling in suggestions, and docc should indicate the deprecation. Renamings are a mild source-break, so having them kick in with the language mode is a good time. We could also decide to bring the deprecations forward at a later time (e.g. Swift 5.19).

Reply:
swift-testing also needs a bit of runway to adopt, because we need to wait for people to have toolchains with the new function before we can safely write:

```swift
#if swift(>=5.11)
useNewInit()
#else
useOldInit()
#endif
```

(We're planning to adopt once various dusts settle.)